### ISSUE TITLE
Help with BIOMOD_xxx - [format for default setting]

### ISSUE THREAD

**Issue by SLL629** [biomodhub/biomod2#401]
Dear biomod Team,

I am new in using Biomod2, I try follow the suggested steps for Biomod2 ver 4.2.5 on 'Help with BIOMOD_xxx - [changing model settings]#340, however I am facing problem while define model options changing compared to OptionBigBoss. In the current study, I would like first to test all algorithms at their default parameters, only then follow by needed parameters adjustments.  I understand that in this step I should use the default setting "names(xxxdata_opt@options$ANN.binary.nnet.nnet@args.values)", "names(xxxdata_opt@options$CTA.binary.rpart.rpart@args.values)"...etc. Nevertheless, I got error when running the BIOMOD_modeling saying "task 2 failed - "task 12 failed - "non-numeric argument to binary operator""", where 5 runs on ANN model and 8 runs on MAXENT model failed.  Below are the script I am using:

## format the data ----
xxxdata_data <- 
  BIOMOD_FormatingData(
    resp.var = xxxdata_occ['xxxdata'],
    resp.xy = xxxdata_occ[, c('longitude', 'latitude')],
    expl.var = bioclim_ZA_sub,
    resp.name = "xxx",
    PA.nb.rep = 2,
    PA.nb.absences = 500,
    PA.strategy = 'random'
  )

## define models options ----
allModels = c('ANN', 'CTA', 'FDA', 'GAM', 'GBM', 'GLM'
              , 'MARS', 'MAXENT', 'MAXNET', 'RF', 'SRE', 'XGBOOST')

xxxdata_opt <- 
  bm_ModelingOptions(
  data.type = 'binary',
  models = allModels,
  strategy = 'default',
  user.val = NULL,
  user.base = "default",
  bm.format = xxxdata_data,
  calib.lines = NULL
)

names(xxxdata_opt@options$ANN.binary.nnet.nnet@args.values)
names(xxxdata_opt@options$CTA.binary.rpart.rpart@args.values)
names(xxxdata_opt@options$FDA.binary.mda.fda@args.values)
names(xxxdata_opt@options$GAM.binary.mgcv.gam@args.values)
names(xxxdata_opt@options$GBM.binary.gbm.gbm@args.values)
names(xxxdata_opt@options$GLM.binary.stats.glm@args.values)
names(xxxdata_opt@options$MARS.binary.earth.earth@args.values)
names(xxxdata_opt@options$MAXENT.binary.MAXENT.MAXENT@args.values)
names(xxxdata_opt@options$MAXNET.binary.maxnet.maxnet@args.values)
names(xxxdata_opt@options$RF.binary.randomForest.randomForest@args.values)
names(xxxdata_opt@options$SRE.binary.biomod2.bm_SRE@args.values)
names(xxxdata_opt@options$XGBOOST.binary.xgboost.xgboost@args.values)

## BIOMOD modeling ---
xxxdata_models <- BIOMOD_Modeling(bm.format = xxxdata_data,
                                    modeling.id = "AllModels",
                                    models = allModels,
                                    CV.strategy = 'random',
                                    CV.nb.rep = 2,
                                    CV.perc = 0.8,
                                    metric.eval = c('TSS','ROC','KAPPA','ACCURACY','BIAS','POD','FAR','POFD','SR','CSI','ETS','HK','HSS','OR','ORSS'),
                                    var.import = 2,
                                    seed.val = 42)

So, here is the output report: 
-=-=-=--=-=-=- xxxdata_PA1_RUN1_ANN 
# weights:  19
initial  value 217.937275 
final  value 203.680982 
converged
	Model scaling...
*** single value predicted
   ! Note :  xxxdata_PA1_RUN1_ANN failed!

-=-=-=--=-=-=- xxxdata_PA1_RUN1_CTA 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_FDA 
	Model scaling...
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_GAM 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_GBM 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_GLM 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_MARS 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_MAXENT 
		Creating Maxent Temp Proj Data...
*** Error in MAXENT, more info available in  ./xxxdata/models/AllModels/xxxdata_PA1_RUN1_MAXENT_outputs/maxent.stderr
*** inherits(g.pred,'try-error')
   ! Note :  xxxdata_PA1_RUN1_MAXENT failed!

-=-=-=--=-=-=- xxxdata_PA1_RUN1_MAXNET 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_RF 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_SRE 
	Evaluating Model stuff...

-=-=-=--=-=-=- xxxdata_PA1_RUN1_XGBOOST 
[1]	train-rmse:0.352489 
[2]	train-rmse:0.249056 
[3]	train-rmse:0.175766 
[4]	train-rmse:0.124277 

	Evaluating Model stuff...
For ANN model, is that means that I can only get it work with optimize parameters? For MaXENT model, I am unsure whether it is because of the path to the JAR file. How should I write the code lines in this case.

Secondly, could anyone show me the format in define model options changing compared to OptionBigBoss. I keep obtained errors at step BIOMOD_modelling  with "Error in BIOMOD_Modeling(bm.format = xxxdata, modeling.id = "AllModels",  : names(OPT.user@options[['ANN.binary.nnet.nnet']]@args.values) must be '_PA1_RUN1', '_PA1_RUN2', '_PA2_RUN1', '_PA2_RUN2', '_allData_allRun', '_PA1_allRun' and '_PA2_allRun".

Appreciate with the suggestions and helps.

Cheers.
Edna

**Comment by HeleneBlt** [comment]
Hello Edna, 

Thanks for reporting 🙏 
Indeed, it seems there is a problem somewhere. 
So I can help you, can you share: 

- the output of `show(xxxdata_data)`
- the error message in ' ./xxxdata/models/AllModels/xxxdata_PA1_RUN1_MAXENT_outputs/maxent.stderr'

Could you also check if a model is missing in the file  './xxxdata/models/AllModels/' ? 


For the other part, here is an example. Let's say I want to change some parameters of the ANN model and MAXENT. 

```
## Define ANN options changing compared to OptionsBigboss 
user.ANN <- list('_allData_allRun' = list(size = 2))

## Define MAXENT options changing compared to OptionsBigboss
user.MAXENT <- list('_allData_allRun' = list(path_to_maxent.jar = getwd()
                                             , product = FALSE
                                             , threshold = FALSE
                                             , hinge = FALSE))

## Gather in one list
## Models names can be found in OptionsBigboss@models
user.val <- list(ANN.binary.nnet.nnet = user.ANN
                 , MAXENT.binary.MAXENT.MAXENT = user.MAXENT)

xxxdata_models <- BIOMOD_Modeling(bm.format = xxxdata_data,
                                    modeling.id = 'AllModels',
                                    models = allModels,
                                    CV.strategy = 'random',
                                    CV.nb.rep = 2,
                                    CV.perc = 0.8,
                                    OPT.user.val = user.val,
                                    OPT.user.base = 'bigboss',
                                    metric.eval = c('TSS','ROC'),
                                    var.import = 2,
                                    seed.val = 42)

```
Please note that until we solve the first problem, this code probably won't work either. I hope it helps a little anyway.

Hélène

**Comment by SLL629** [comment]
hi Hélène,

Sorry for late response.
I here attached the files on "the output of show(xxxdata_data)", and "maxent.stderr" for your reference.

Within the Allmodel directory, there are 7 folders namely xxxdata_allData_allRun_MAXENT_output, xxxdata_PA1_allRun_MAXENT_outputs, xxxdata_PA1_RUN1_MAXENT_outputs, xxxdata_PA1_RUN2_MAXENT_outputs, xxxdata_PA2_allRun_MAXENT_outputs, xxxdata_PA2_RUN1_MAXENT_outputs, xxxdata_PA2_RUN2_MAXENT_outputs. 

Many thanks!

Edna

![Biomod2](https://github.com/biomodhub/biomod2/assets/156922694/8df14b1f-e932-4ed3-bc6b-dec388d904b4)
![allModelmaxentstderrerror](https://github.com/biomodhub/biomod2/assets/156922694/dd07d5b8-3e9b-4f91-b064-d8ef6fe234bb)

**Comment by HeleneBlt** [comment]
Hi Edna, 

So for MAXENT, the error doesn't come from Biomod2 but probably from java. You can find some solutions [here](https://github.com/microsoft/WSL/issues/7006) or [here](https://stackoverflow.com/questions/10165761/java-cant-connect-to-x11-window-server-using-localhost10-0-as-the-value-of-t). It seems to be a common issue, you must find different solutions. 

For ANN, individual models may fail for numerous reasons. For instance, it can happen because you have a limited amount of data or lack of contrast in environmental data. There may also be some random components in the different methods so if you re-run `BIOMOD_Modeling` you may see that some models that previously failed now work properly. And yes, if you optimize the parameters, it will probably work. 

Also, as we pushed several commits, I will invite you to update your version of Biomod.2-5. Error messages can be clearer.

Let me know if you find more errors.
Hélène

**Comment by SLL629** [comment]
Hi Hélène,

I tried to test only on one algorithm (e.g. XGBOOST), at the i got the error on "Evaluating Model stuff...Error in { : 
  task 1 failed - "task 12 failed - "non-numeric argument to binary operator"". I attached the image FYI. How should I solve this error?

In addition to the java issue, I wonder if we can save the result on maxent run directly with a code line during the 'BIOMOD_Modeling' step?

Many thanks.

Edna
![biomodmodellingerror](https://github.com/biomodhub/biomod2/assets/156922694/13e2540b-945a-4cc5-b2fa-9538eaf1661d)

**Comment by HeleneBlt** [comment]
Hi Edna, 

Sorry for my late answer 😬  
If you have the same error for all the models, it must come from the data. Do you have some factors in your explanation variables? If possible, can you send me your data (by email if you want helene.blancheteau[at]univ-grenoble-alpes.fr) ? 

I'm not sure I understand your question about the result of `Maxent,` do you want to save the error message somewhere else? 

Hélène
